cng/ecdh.go

```diff
@@ -0,0 +1,206 @@
+		return
+	}
+	v, err := loadOrStoreAlg(id, bcrypt.ALG_NONE_FLAG, "", func(h bcrypt.ALG_HANDLE) (interface{}, error) {
+		return ecdhAlgorithm{h, id}, nil
+	})
+	if err != nil {
+		return ecdhAlgorithm{}, 0, err
+	}
+	return v.(ecdhAlgorithm), bits, nil
+}
+
+type PublicKeyECDH struct {
+	hkey  bcrypt.KEY_HANDLE
+	bytes []byte
+}
+
+func (k *PublicKeyECDH) finalize() {
+	bcrypt.DestroyKey(k.hkey)
+}
+
+type PrivateKeyECDH struct {
+	hkey bcrypt.KEY_HANDLE
+}
+
+func (k *PrivateKeyECDH) finalize() {
+	bcrypt.DestroyKey(k.hkey)
+}
+
+func ECDH(priv *PrivateKeyECDH, pub *PublicKeyECDH) ([]byte, error) {
+	// First establish the shared secret.
+	var secret bcrypt.SECRET_HANDLE
+	err := bcrypt.SecretAgreement(priv.hkey, pub.hkey, &secret, 0)
+	if err != nil {
+		return nil, err
+	}
+	defer bcrypt.DestroySecret(secret)
+
+	// Then we need to export the raw shared secret from the secret opaque handler.
+	// The only way to do it is using BCryptDeriveKey with BCRYPT_KDF_RAW_SECRET as key derivation function (KDF).
+	// Unfortunately, this KDF is supported starting from Windows 10.
+	kdf := utf16PtrFromString(bcrypt.KDF_RAW_SECRET)
+	var size uint32
+	err = bcrypt.DeriveKey(secret, kdf, nil, nil, &size, 0)
+	if err != nil {
+		return nil, err
+	}
+	agreedSecret := make([]byte, size)
+	err = bcrypt.DeriveKey(secret, kdf, nil, agreedSecret, &size, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	// The raw shared secret is little-endian but Go expects big-endian.
+	// Reverse the slice in-place.
+	inputMid := size / 2
+	for i := uint32(0); i < inputMid; i++ {
+		j := size - i - 1
+		agreedSecret[i], agreedSecret[j] = agreedSecret[j], agreedSecret[i]
+	}
+	runtime.KeepAlive(priv)
+	runtime.KeepAlive(pub)
+	return agreedSecret, nil
+}
+
+func GenerateKeyECDH(curve string) (*PrivateKeyECDH, []byte, error) {
+	h, bits, err := loadEcdh(curve)
+	if err != nil {
+		return nil, nil, err
+	}
+	var hkey bcrypt.KEY_HANDLE
+	err = bcrypt.GenerateKeyPair(h.handle, &hkey, bits, 0)
+	if err != nil {
+		return nil, nil, err
+	}
+	// The key cannot be used until BCryptFinalizeKeyPair has been called.
+	err = bcrypt.FinalizeKeyPair(hkey, 0)
+	if err != nil {
+		bcrypt.DestroyKey(hkey)
+		return nil, nil, err
+	}
+
+	// GenerateKeyECDH returns the public key as as byte slice.
+	// To get it we need to export the raw CNG key blob
+	// and prepend the encoding prefix.
+	_, blob, err := exportCCKey(hkey, false)
+	if err != nil {
+		bcrypt.DestroyKey(hkey)
+		return nil, nil, err
+	}
+	k := new(PrivateKeyECDH)
+	k.hkey = hkey
+	runtime.SetFinalizer(k, (*PrivateKeyECDH).finalize)
+	bytes := append([]byte{ecdhUncompressedPrefix}, blob...)
+	return k, bytes, nil
+}
+
+func NewPrivateKeyECDH(curve string, key []byte) (*PrivateKeyECDH, error) {
+	h, bits, err := loadEcdh(curve)
+	if err != nil {
+		return nil, err
+	}
+	keySize := int(bits+7) / 8
+	if len(key) != keySize {
+		return nil, errInvalidPrivateKey
+	}
+	// zero has enough size to fit P-521 curves.
+	var zero [66]byte
```

Comment:
It could live at package scope, but I like it to be defined just before it is used, else one could be tempted to use it for another algorithm that has other size requirements.

I'll improve the comment and use `521 / 8` 😄